File: src/network/heartbeat.py
"""Heartbeat v0.1 骨架（单连接心跳状态机）。

设计真源：``docs/web-network-layer-v0.1/02-module-breakdown.md``
``Heartbeat 设计`` 节。

核心约束：

- **一个 ``Heartbeat`` 实例 = 一条连接的心跳状态机**，不是全局单例；
  ``NetworkManager`` 持有 ``dict[conn_id, Heartbeat]``，N 连接 = N 个
  实例并存，互不串扰
- ``Heartbeat`` 类**纯被动**：不知道底层 socket、不知道 channel 类型，
  只通过 ``HeartbeatHooks`` 与外界交互
- 多实例并发安全是核心约束；任何共享可变状态都必须放在实例字段内，
  类作用域 / 模块作用域**不允许**出现可变状态

v0.1 后端简化：后端是 server 角色（被动响应 client 发的 ping），所以
本版本 ``Heartbeat`` 只暴露 ``handle_inbound_frame(frame) -> dict | None``
处理 inbound ping 并返回应回的 pong，不需要主动发 ping。接口仍预留
双向能力，v0.3 升级到 server-push ping 时不用动 client 端。

.. note::

   v0.1 step 1 阶段所有方法 body 是 ``raise NotImplementedError``，
   行为实现在 step 3（Heartbeat 状态机）补齐。
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Protocol


class HeartbeatHooks(Protocol):
    """``Heartbeat`` 与外界交互的回调接口。

    v0.1 后端 server 角色只需要两个 hook：``is_open`` 判断底层 socket
    是否可写、``send_pong`` 发回 pong 帧；``send_ping`` / ``on_latency``
    等主动探测语义留给 v0.3 server-push 升级时再加，本版本不引入避免
    Protocol 接口漂移。
    """

    def is_open(self) -> bool:
        """返回底层 socket 是否仍可写。"""
        ...

    async def send_pong(self, frame: dict[str, Any]) -> None:
        """把一条 pong 帧写回 socket。

        :param frame: 已构造好的 pong 帧 dict（包含 ``kind`` /
            ``timestamp_ms`` / ``ts`` 等字段，schema 见
            ``src/web/protocol/ws_frames.py::PongFrame``）
        """
        ...


@dataclass
class HeartbeatConfig:
    """心跳参数；**不设默认值**，必须由调用方传入。

    强制走 ``cfg.web.ws_heartbeat_*`` 三件套真源（已存在），不在
    ``network/`` 层立新常量；所有频道未来共用同一份配置（"所有通道
    公用一个倒计时配置"原则，spec 04 明确禁止覆盖）。

    :ivar interval_ms: 期望相邻两次 ping 的间隔毫秒数；server 角色
        v0.1 不主动发 ping，本字段当前仅作 metadata 备 v0.3 用
    :ivar timeout_ms: 单次 pong 等待超时毫秒数；同上 v0.1 仅 metadata
    :ivar max_missed: 连续未回的 pong 次数上限；同上 v0.1 仅 metadata
    """

    interval_ms: int
    timeout_ms: int
    max_missed: int


class Heartbeat:
    """单连接心跳状态机；``NetworkManager`` per-connection 实例化。

    实例字段全部私有（``_`` 前缀）；多实例并发安全靠"每实例独立持有
    可变状态"实现，不依赖全局锁。
    """

    def __init__(self, config: HeartbeatConfig, hooks: HeartbeatHooks) -> None:
        self._config: HeartbeatConfig = config
        self._hooks: HeartbeatHooks = hooks
        self._missed_pongs: int = 0
        self._pending_ping_ts: int | None = None
        self._latency_ms: int | None = None
        self._started: bool = False

    async def handle_inbound_frame(self, frame: dict[str, Any]) -> dict[str, Any] | None:
        """处理一条入站帧；若为 ping 则返回应回的 pong 帧 dict。

        非心跳相关帧（``kind != "ping"``）返回 ``None``，让上层把帧路由给
        业务 handler。

        服务端 v0.1 角色是 **被动响应**：客户端按 ``intervalMs`` 周期发 ping，
        本方法识别后原样 echo 客户端的 ``ts`` 用于 RTT 计算，并附带服务端
        当前 wall-clock ``timestamp_ms`` 作为信息字段（前端不参与 RTT，仅
        用于服务端时间漂移诊断）。pong 字段 schema 兼容
        :class:`web.protocol.ws_frames.PongFrame`（多出的 ``timestamp_ms``
        被 Pydantic ``extra`` 默认忽略）。

        纯函数式实现，多实例并发安全：不读写任何实例可变状态，所有状态
        来自入参 ``frame``，副作用为 ``None``（pong 发送由上层 ``send_pong``
        hook 负责）。

        :param frame: 已被上层解析为 ``dict`` 的入站帧
        :returns: 若 ``frame["kind"] == "ping"``，返回 pong 帧 dict
            ``{"kind": "pong", "ts": <原 ts>, "timestamp_ms": <服务端 ms>}``；
            否则 ``None``
        """
        if not isinstance(frame, dict):
            return None
        if frame.get("kind") != "ping":
            return None
        return {
            "kind": "pong",
            "ts": frame.get("ts"),
            "timestamp_ms": int(time.time() * 1000),
        }

    @property
    def latency_ms(self) -> int | None:
        """最近一次 RTT；v0.1 后端 server 角色不主动探测，恒为 ``None``。"""
        return self._latency_ms

File: web/src/network/heartbeat.ts
/**
 * kongming-agent web-network-layer v0.1 · Heartbeat 前端实现
 *
 * ## 这份文件是什么
 *
 * **单连接心跳状态机**——一个 `Heartbeat` 实例对应一条 socket。`NetworkManager`
 * 维护 `Map<connId, Heartbeat>`，N 条连接 = N 个并发 Heartbeat 实例。
 *
 * 设计文档：
 * - `docs/web-network-layer-v0.1/02-module-breakdown.md`（接口草图）
 * - `docs/web-network-layer-v0.1/04-data-and-state.md`（内部字段 + HeartbeatConfig 真源）
 *
 * ## 核心约束
 *
 * - **多实例并发安全**：每个实例自管 timer 句柄 / 计数器，互不污染。
 * - **纯被动设计**：Heartbeat **不持有** socket 引用、**不知道** channel 类型，
 *   只通过 `HeartbeatHooks` 4 个回调与外界交互。
 * - **数值真源**：`HeartbeatConfig` 由 manager 注入来自 `useHeartbeatConfig` 的真值，
 *   本模块**严禁**立默认常量。
 */

// ---------------------------------------------------------------------------
// 公共类型
// ---------------------------------------------------------------------------

/**
 * Heartbeat 与外界交互的 4 个回调。
 *
 * - `isOpen`            —— 底层 socket 是否可写（重连过渡期可能 false）
 * - `sendPing`          —— 发一帧 ping；ts 由 Heartbeat 生成传入
 * - `closeForReconnect` —— 连续丢 `maxMissed` 次时调；由 manager 触发重连
 * - `onLatency`         —— RTT 回调（可选）；handlePong 时调，参数为 ms
 */
export interface HeartbeatHooks {
  isOpen(): boolean;
  sendPing(ts: number): void;
  closeForReconnect(): void;
  onLatency?(ms: number): void;
}

/**
 * Heartbeat 三参数；**不设默认值**，必须由 manager 从 `useHeartbeatConfig` 注入。
 *
 * 真源：`cfg.web.ws_heartbeat_interval_ms` / `ws_heartbeat_timeout_ms` /
 * `ws_heartbeat_max_missed`。
 */
export interface HeartbeatConfig {
  /** 前台 ping 间隔，毫秒（来源 cfg.web.ws_heartbeat_interval_ms，默认 30000） */
  intervalMs: number;
  /**
   * 后台 ping 间隔，毫秒（来源 cfg.web.ws_heartbeat_background_interval_ms，默认 60000）。
   *
   * 浏览器 tab 切到后台时切换为此间隔。Chrome 后台节流 setInterval 到 ≥1min，
   * 前台 30s 在后台变成噪音 + RTT 测出来是节流延迟不是真网络 RTT。
   * 用更稀疏的间隔减少误导；切回前台时切回 `intervalMs` + 立即 probe。
   */
  backgroundIntervalMs: number;
  /** 单次 pong 超时，毫秒（来源 cfg.web.ws_heartbeat_timeout_ms） */
  timeoutMs: number;
  /** 连续丢几次判死 → closeForReconnect（来源 cfg.web.ws_heartbeat_max_missed） */
  maxMissed: number;
}

// ---------------------------------------------------------------------------
// Heartbeat
// ---------------------------------------------------------------------------

/**
 * 单连接心跳状态机。NetworkManager 在每次新建 socket 时 `new Heartbeat(config, hooks)`，
 * socket open 后调 `start()`，close 时调 `stop()`，收到 pong 帧时调 `handlePong(ts)`。
 */
export class Heartbeat {
  private readonly _config: HeartbeatConfig;
  private readonly _hooks: HeartbeatHooks;

  /** setInterval 句柄；null 表示 stop 状态 */
  private _heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  /** 单次 pong 超时 setTimeout 句柄；null 表示当前没 pending ping */
  private _pongTimer: ReturnType<typeof setTimeout> | null = null;
  /** 连续未回的 pong 次数 */
  private _missedPongs = 0;
  /** 最近一次未确认 ping 的时间戳（用于 RTT 计算回填校验） */
  private _pendingPingTs: number | null = null;
  /** 最近一次 RTT；尚未握过手时为 null */
  private _latencyMs: number | null = null;
  /** 状态机是否在跑（防 start 重入） */
  private _started = false;
  /** 当前 setInterval 用的 ms 值；setBackgrounded 切换前台/后台时变 */
  private _currentIntervalMs: number;

  constructor(config: HeartbeatConfig, hooks: HeartbeatHooks) {
    this._config = config;
    this._hooks = hooks;
    this._currentIntervalMs = config.intervalMs;
  }

  /**
   * 启动心跳：注册 `setInterval(intervalMs)` 周期发 ping。
   *
   * 幂等：已 started 的实例再调 start 直接返回，不会双注册 timer。
   */
  start(): void {
    if (this._started) return;
    this._started = true;
    // 防御：构造期到 start 之间外部可能复用实例，清掉残留
    this._clearTimers();
    this._missedPongs = 0;
    this._pendingPingTs = null;
    this._heartbeatTimer = setInterval(() => {
      this._sendPing();
    }, this._currentIntervalMs);
  }

  /**
   * 切换前台 / 后台模式：使用对应 interval 重启 setInterval。
   *
   * - `hidden=true`：用 `backgroundIntervalMs`（默认 60s，更稀疏减少节流噪音）
   * - `hidden=false`：用 `intervalMs`（默认 30s，前台正常频率）
   *
   * 若当前未 start（stop 状态）→ 只更新 `_currentIntervalMs`，下次 start 生效。
   * 若当前已 start → 立刻重启 setInterval 切到新 interval；正在 pending 的 pong
   * 超时计时不变（避免重启时把刚发的 ping 状态清掉）。
   *
   * @param hidden 来自 `document.hidden`：true=切到后台，false=切回前台
   */
  setBackgrounded(hidden: boolean): void {
    const nextInterval = hidden
      ? this._config.backgroundIntervalMs
      : this._config.intervalMs;
    if (nextInterval === this._currentIntervalMs) return;
    this._currentIntervalMs = nextInterval;
    if (!this._started) return;
    // 重启 setInterval 切到新频率；不动 _pongTimer / _pendingPingTs / _missedPongs
    if (this._heartbeatTimer !== null) {
      clearInterval(this._heartbeatTimer);
    }
    this._heartbeatTimer = setInterval(() => {
      this._sendPing();
    }, this._currentIntervalMs);
  }

  /**
   * 停止心跳：清理 interval / pong-timeout / 计数器 / pending ts。
   *
   * 幂等；socket close / channel close 时调。
   */
  stop(): void {
    this._started = false;
    this._clearTimers();
    this._missedPongs = 0;
    this._pendingPingTs = null;
  }

  /**
   * 收到 pong 帧时调用。
   *
   * - 清 pong 超时 timer
   * - 重置 `_missedPongs = 0`
   * - 计算 RTT = `Date.now() - ts` → 写 `_latencyMs` + 调 `hooks.onLatency?(rtt)`
   *
   * @param ts 此 pong 对应的 ping 时间戳（server 原样 echo 回来）
   */
  handlePong(ts: number): void {
    // 收到任意 pong 都视作 server 端活着；清掉 pong 超时 timer
    if (this._pongTimer !== null) {
      clearTimeout(this._pongTimer);
      this._pongTimer = null;
    }
    this._missedPongs = 0;
    // 如果 ts 跟最近一次 pending ping 不匹配，仍接受（server echo 可能延迟若干轮）
    // 这里读字段一次，避免 TS noUnusedLocals 抱怨；不强制相等。
    const expected = this._pendingPingTs;
    if (expected !== null && expected > ts) {
      // 防御性日志：server echo 比客户端发出还早 → 时钟问题
      console.warn(
        "[Heartbeat] pong ts < pending ts; clock skew?",
        { expected, ts },
      );
    }
    this._pendingPingTs = null;
    const rtt = Date.now() - ts;
    // 防御：负 RTT（时钟回拨 / ts 篡改）截到 0；超大 RTT 仍如实上报供诊断
    this._latencyMs = rtt < 0 ? 0 : rtt;
    if (this._hooks.onLatency) {
      try {
        this._hooks.onLatency(this._latencyMs);
      } catch (err) {
        // listener 抛错不应传染心跳；上层 manager 可能有 zustand setter 之类，
        // 测试环境也可能拿 onLatency 抛错验证健壮性
        console.error("[Heartbeat] onLatency listener threw", err);
      }
    }
  }

  /**
   * 立即发一帧 ping（不等下一个 interval）。
   *
   * 用途：visibilitychange 切回前台时 manager 对所有 Heartbeat 调 `probe()`，
   * 立即探测连接是否还活着。stop 状态下 probe 是 no-op。
   */
  probe(): void {
    if (!this._started) return;
    this._sendPing();
  }

  /**
   * 最近一次 RTT，毫秒；尚未握过手 / 已 stop 时返回 null。
   */
  get latencyMs(): number | null {
    return this._latencyMs;
  }

  // -------------------------------------------------------------------------
  // 内部实现
  // -------------------------------------------------------------------------

  /**
   * 发一帧 ping + 注册 pong 超时 timer。
   *
   * - socket 不 open 时直接 skip（不增 missed 计数；让重连流程接管判定）
   * - 已有 pendingPingTs 时不再发新 ping，避免双重 timer 叠加
   */
  private _sendPing(): void {
    if (!this._hooks.isOpen()) return;
    if (this._pongTimer !== null) {
      // 上一帧 ping 的 pong 超时未到，先不发新 ping（避免双 timer 触发误判死）
      return;
    }
    const ts = Date.now();
    this._pendingPingTs = ts;
    try {
      this._hooks.sendPing(ts);
    } catch (err) {
      // 底层 send 失败（如 readyState 异常）：清状态，让重连接管
      console.error("[Heartbeat] sendPing threw", err);
      this._pendingPingTs = null;
      return;
    }
    this._pongTimer = setTimeout(() => {
      this._onPongTimeout();
    }, this._config.timeoutMs);
  }

  /**
   * 单次 pong 超时回调：累加 missedPongs；达到上限调 closeForReconnect。
   */
  private _onPongTimeout(): void {
    this._pongTimer = null;
    this._pendingPingTs = null;
    this._missedPongs += 1;
    if (this._missedPongs >= this._config.maxMissed) {
      try {
        this._hooks.closeForReconnect();
      } catch (err) {
        console.error("[Heartbeat] closeForReconnect threw", err);
      }
      // 由 manager 在 onclose 里调 stop()；这里不主动 stop 防 stop 抢先把
      // _missedPongs 清零导致后续诊断丢失。
    }
  }

  private _clearTimers(): void {
    if (this._heartbeatTimer !== null) {
      clearInterval(this._heartbeatTimer);
      this._heartbeatTimer = null;
    }
    if (this._pongTimer !== null) {
      clearTimeout(this._pongTimer);
      this._pongTimer = null;
    }
  }
}
